dma: stm32 driver cleanup

erwango opened the pull request:
This change is an extract of #23157 focused on changes that do not impact SPI

EDIT by avisconti::
Tested on shield iks02a1 with microphone sample
(https://github.com/avisconti/zephyr/tree/test_new_dmastm32)

FRASTM committed:
drivers: dma: include duplicated

remove duplicates line

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: bindings: dma controller for stm32 soc series

Complete documentation for STM32 DMA binding

Signed-off-by: Francois Ramu <francois.ramu@st.com>
Signed-off-by: Erwan Gouiou <erwan.gouriou@linaro.org>

FRASTM committed:
drivers: dma: initialize peripheral request dma on stm32

Set the peripheral ID for the Channel Selection register

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
include: bindings: dma define the bindings for stm32 soc series

This defines the configuration for the DMA client
as specifed in the ./dts/bindings/dma/st,stm32-dma.yaml

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: dma: irq handler of the dma_stm32

define the irq handler in case of DMA V1 or V2,
for the stm32x soc series
with DMA V1 raise Fifo Error if enabled
with V2, handle the Global Interrupt

Signed-off-by: Francois Ramu <francois.ramu@st.com>
Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

FRASTM committed:
drivers: dma: control of null buffer address for stm32 dma transfers

Raise warning if the buffer address of the source or the dest is NULL

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: dma: channel direction of the dma_stm32

Return an error if the direction of the channel is wrong

Signed-off-by: Francois Ramu <francois.ramu@st.com>
Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

FRASTM committed:
drivers: dma: clean up header

Move tables declaration as they are only used locally for now

Signed-off-by: Francois Ramu <francois.ramu@st.com>
Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

cybertale commented on include/dt-bindings/dma/stm32_dma.h:
```diff
@@ -1,16 +1,38 @@
+/* perih data size defined on bits 11-12 */
+/* 0 -> 1 byte, 1 -> 2 bytes, 2 -> 4 bytes */
+#define STM32_DMA_CONFIG_PERIPH_DATA_SIZE(config)	\
+	((((config >> 11) & 0x3) == 0) ? 1 : (((config >> 11) & 0x3) * 2))
```
I think
```
#define STM32_DMA_CONFIG_PERIPH_DATA_SIZE(config) (1 << ((config >> 11) & 0x3))
```
is better, because this calculation is just `2^((config >> 11) & 0x3)`. What do you think?

cybertale commented on include/dt-bindings/dma/stm32_dma.h:
```diff
@@ -1,16 +1,38 @@
+/* memory data size defined on bits 13, 14 */
+/* 0 -> 1 byte, 1 -> 2 bytes, 2 -> 4 bytes */
+#define STM32_DMA_CONFIG_MEM_DATA_SIZE(config)		\
+	((((config >> 13) & 0x3) == 0) ? 1 : (((config >> 13) & 0x3) * 2))
```
Same as above.

cybertale commented on drivers/dma/dma_stm32.c:
```diff
@@ -198,20 +201,26 @@ static int dma_stm32_get_periph_increment(enum dma_addr_adj increment,
 static int dma_stm32_configure(struct device *dev, u32_t id,
 			       struct dma_config *config)
 {
+
```
Why this empty line?

FRASTM replied:
no reason, removed

cybertale commented on drivers/dma/dma_stm32.c:
```diff
@@ -198,20 +201,26 @@ static int dma_stm32_get_periph_increment(enum dma_addr_adj increment,
 	const struct dma_stm32_config *dev_config =
 					dev->config->config_info;
 	DMA_TypeDef *dma = (DMA_TypeDef *)dev_config->base;
+	struct dma_stm32_stream *stream = &data->streams[id - STREAM_OFFSET];
```
Out of curiosity, why are we moving this line? General idea of initializations should be following the reverse christmas tree format, which means the long lines should go up rather than down.

FRASTM replied:
Ok, (the reason is that line has been moved several time,)
Restoring initial position

cybertale commented on drivers/dma/dma_stm32.c:
```diff
@@ -198,20 +201,26 @@ static int dma_stm32_get_periph_increment(enum dma_addr_adj increment,
+	id = id - STREAM_OFFSET;
+
 	if (id >= data->max_streams) {
+		LOG_ERR("cannot configure the dma stream %d.", id);
```
Maybe here we use "channel index setup is not valid with value %d" is better?

FRASTM replied:
OK, The fact is that this "Id" refer to a dma stream. So I propose 
LOG_ERR("Channel %d is not valid for this dma.", id);

cybertale commented on drivers/dma/dma_stm32.c:
```diff
@@ -304,7 +313,6 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
 		periph_addr_adj = config->head_block->dest_addr_adj;
 		break;
 	/* Direction has been asserted in dma_stm32_get_direction. */
-	}
```
Is this a typo?

FRASTM replied:
yes, fixed elsewhere, now fixed in this commit

erwango commented:
@cybertale nice to see you back, hope you are doing fine.
Could you also have a look to #23157 and #23155 which are 2 functional improvements we would like to see progressing?. This PR is essentially an abstract of these as preparation work.

cybertale commented:
@erwango , thanks for the concern! Due to the reason that we all know, the COVID-19 pandemic, we still can not go back to school, where all my hardwares are. So I am not able to do any actually tests. But I do can help review the code. I'll look into PRs you mentioned.

erwango committed:
drivers: dma: channel configuration for dma_stm32

Control values when configuring the dma channel
According to the soc specification,
the dma V2 channel counts from 1.
the dma V1 stream counts from 0.

Signed-off-by: Francois Ramu <francois.ramu@st.com>
Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@cybertale, I think I've fixed all your comments

FRASTM commented:
I have introduced the dma: stm32 driver cleanup #24333 
as an equivalent PR

FRASTM commented on include/dt-bindings/dma/stm32_dma.h:
```diff
@@ -1,16 +1,38 @@
+/* periph increment defined on bit 9 as true/false */
+#define STM32_DMA_CONFIG_PERIPHERAL_ADDR_INC(config)	((config >> 9) & 0x1)
+/* mem increment defined on bit 10 as true/false */
+#define STM32_DMA_CONFIG_MEMEMORY_ADDR_INC(config)	((config >> 10) & 0x1)
```
STM32_DMA_CONFIG_MEMORY_ADDR_INC

erwango replied:
oups, txs!

FRASTM reviewed (commented):
commit msg 
"Move tables declaration **as** the are only used locally for now"

erwango commented:
>commit msg
"Move tables declaration as the are only used locally for now"

Fixed txs.

avisconti commented on include/dt-bindings/dma/stm32_dma.h:
```diff
@@ -1,16 +1,38 @@
+						(1 << ((config >> 13) & 0x3))
+/* priority increment offset defined on bit 15 */
+#define STM32_DMA_CONFIG_PERIPHERAL_INC_FIXED(config)	((config >> 15) & 0x1)
+/* priority defined on bits 16-17 as 0, 1, 2, 3 */
```
Seems like before there was a bug, isn't it?
Or maybe the semantic of the defines is changed?

FRASTM replied:
yes or No, I did not find where this is used. This field is valid for V1 only.
Ok, I let it unchanged

avisconti replied:
I prefer your new way! It seems cleaner.

avisconti replied:
So I agree with your changes.

FRASTM replied:
do you mean 
`#define STM32_DMA_CONFIG_PERIPHERAL_INC_FIXED(config)	((config >> 15) & 0x1)` 

or 
```
#ifdef CONFIG_DMA_STM32_V1
/* priority increment offset defined on bit 15 */
#define STM32_DMA_CONFIG_PERIPH_INC_FIXED(config)	(config & 0x1 << 15)
#endif /* CONFIG_DMA_STM32_V1 */
```

avisconti replied:
I mean the complete block of changes with the ((config >> 15) & 0x1) style.

FRASTM replied:
ok

avisconti commented on drivers/dma/dma_stm32.h:
```diff
@@ -45,6 +45,15 @@ struct dma_stm32_config {
+/* so decrease to set range from 0 from now on */
+#define STREAM_OFFSET 1
+#endif /* CONFIG_DMA_STM32_V1 */
+
```
It seems to me that for L4 is 1 to 7 (not 8). Correct?
For F4 is 0 to 7.

FRASTM replied:
Yes, this is very confusing in the stm32 ref manuals.
For F4 (DMA V1) requests are called 'channels' 
For L4 (DMA V2) channels (what DMA V1 calls 'streams') have several requests.
I tried to format in the bindings the channel (or stream for DMA_V1) in the first position and the request (or channel in DMA_V1) in the second position, based on the table given by the ref Man ( whatever the numbering from 0 to 7 or from 1 to 8 is).
For example on L476:
 the SPI1_TX  on Channel 3 (from 1 to 8) has request 1 (from 0 to 7)
 the SPI1_RX on Channel 2 (from 1 to 8) has request 1 (from 0 to 7) 
For example on F411:
 the SPI1_TX  on stream 5 (from 0 to 7) has request channel 3 (from 0 to 7)  
 the SPI1_RX on stream 2 (from 0 to 7) has request channel  3 (from 0 to 7)

avisconti replied:
YEs, very confusing.

avisconti replied:
OK, I think it is correct in this way.

avisconti commented on drivers/dma/dma_stm32_v2.c:
```diff
@@ -117,7 +117,7 @@ static void (*func_ll_clear_te[])(DMA_TypeDef *DMAx) = {
 };
 
-static void (*func_ll_clear_gi[])(DMA_TypeDef *DMAx) = {
+void (*func_ll_clear_gi[])(DMA_TypeDef *DMAx) = {
```
Why did you remove the static here and declared it extern?
I understand that it is gonna used only in this file. So why?

FRASTM replied:
func_ll_clear_gi() function is used in the dma_stm32_irq_handler() of the dma_stm32.c

avisconti replied:
OK, I missed that one! Sorry.

avisconti commented on drivers/dma/dma_stm32.c:
```diff
@@ -287,6 +287,16 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
 	stream->src_size	= config->source_data_size;
 	stream->dst_size	= config->dest_data_size;
 
+	if ((&config->head_block->source_address == NULL)
```
What is the meaning of this check? Test if config->head_block is NULL? So, why not checking it in this way?
Moreover, if this is what you want to check, maybe is better to test it in the beginning of the routine.

FRASTM replied:
We control here that the buffer for receiving or sending data is not 0.
 In case it is 0, it means the transfer will read (or write) mem address 0x0000 0000, which is not actually an error in terms of DMA memory transfer. 
Yes, this can be moved above.

avisconti replied:
Wait a moment.
I think that &config->head_block->source_address is always != from 0, isn't it?
THis is the address of a field in a structure. so != from 0.
I guess that check may be only:
if (config->head_block->source_address != 0)

What do you think?

FRASTM replied:
That was the purpose of the testcase: `spi_null_tx_buf` in tests/drivers/spi/spi_loopback
to assign set 0 for the transmit (origin) mem address. and the structure tx_bufs is valid.
Here is what I want to check.

avisconti replied:
OK, so do you agree that if (config->head_block->source_address != 0) should be enough?

FRASTM replied:
yes, agreed

avisconti commented on drivers/dma/dma_stm32.c:
```diff
@@ -324,6 +324,10 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
 		break;
 	/* Direction has been asserted in dma_stm32_get_direction. */
+	default:
+		LOG_ERR("Channel direction error. %d",
```
maybe you may want to get rid of the '.' here.

FRASTM replied:
changed to` LOG_ERR("Channel direction error (%d)."`

avisconti replied:
ack

erwango commented:
@avisconti, I've updated the PR with the 2 agreed changes

avisconti reviewed (approved):
LGTM

galak commented on include/dt-bindings/dma/stm32_dma.h:
```diff
@@ -1,16 +1,38 @@
 #define STM32_DMA_CONFIG_PRIORITY(config)		((config >> 16) & 0x3)
 
 /* macros for features */
+#ifdef CONFIG_DMA_STM32_V1
```
We should avoid CONFIG defines in this header if at all possible.

erwango replied:
ok, this can be added.

erwango replied:
Removed the conditional part.
Difference is already dealt with inside the driver.

galak commented on dts/bindings/dma/st,stm32-dma.yaml:
```diff
@@ -23,15 +23,67 @@ properties:
+       	       0x2: high
+       	       0x3: very high
+    4. features: A 32bit bitfield value specifying DMA features
+        -bit 0-1 or not used: DMA FIFO threshold selection
```
Not sure I follow here.  What does 'or not used' mean?

erwango replied:
Depending on the IP, it might not be used (this is the case for _v1)

erwango replied:
This was actually not used in V2 case. Added a clarification on that.

galak reviewed (commented):
a few things around Device tree changes.

erwango commented:
@galak, this should be ok now
